JSIL.Libraries/Includes/Core/Types/Classes/System.Array.js

```diff
@@ -1,129 +1,400 @@
 ﻿JSIL.ImplementExternals("System.Array", function ($) {
-    $.RawMethod(true, "CheckType", JSIL.IsSystemArray);
+  $.RawMethod(true, "CheckType", JSIL.IsSystemArray);
 
-    $.RawMethod(true, "Of", function Array_Of() {
-        // Ensure System.Array is initialized.
-        var _unused = $jsilcore.System.Array.Of;
+  $.RawMethod(true, "Of", function Array_Of() {
+    // Ensure System.Array is initialized.
+    var _unused = $jsilcore.System.Array.Of;
 
-        return $jsilcore.ArrayOf.apply(null, arguments);
-    });
+    return $jsilcore.ArrayOf.apply(null, arguments);
+  });
 });
 
 JSIL.MakeClass("System.Object", "System.Array", true, [], function ($) {
-    $.SetValue("__IsArray__", true);
+  $.SetValue("__IsArray__", true);
 
-    $.RawMethod(false, "GetLength", function () {
-        return this.length;
-    });
-    $.RawMethod(false, "GetLowerBound", function () {
-        return 0;
-    });
-    $.RawMethod(false, "GetUpperBound", function () {
-        return this.length - 1;
+  $.RawMethod(false, "GetLength", function (dimension) {
+    if (this.LowerBounds || false) {
```

Comment:
Personally, I don't like very much name Multidimensional array, as it can be really single-dimension (nut not zero-based, or even zero based, if somebody really wants it).
In ECMA-335 they named them as Array and Vector.
Unfortunately, in C# (and in JSIL java part) we really name use name Array instead Vector and Multidimensional array instead of Array.